### PR TITLE
[docs] Add dynamic scrolling page navigation

### PR DESCRIPTION
**Pull request by barthap**
> # **Closed in favour of #10230**

## Why

Some documentation pages are very long (e.g. [Notifications](https://docs.expo.io/versions/v38.0.0/sdk/notifications/), [app.json docs](https://docs.expo.io/versions/v38.0.0/config/app/)) and difficult to navigate with. Besides that, RN docs have this feature (look at [this right sidebar](https://reactnative.dev/docs/flatlist)).

## How

Inspired by RN docs, added right sidebar with dynamic page navigation/table of contents with heading tree.

### Some technical details

> TL;DR - I heading management with current permalink generation. Created right sidebar is almost a twin for the left one.

_WIP - this section may change!_

<details><summary>How it works under the hood (detailed writeup, click to expand)</summary>

Each page goes through several processing steps before it can be displayed in the browser:
1. Markdown `.md` file is loaded by Webpack.
2. It is processed by our custom webpack loader `~/common/md-loader.js` which extracts document metadata and wraps the content with HOC `withDocumentationElements`. The metadata is saved as `export const meta = {}` statement.
3. It is processed by `mdx-js` webpack loader, which transforms markdown tags into HTML/JSX tags. This loader accepts two types of plugins: _remark_ and _rehype_ which allows modifications at two different processing stages.
4. Now the React/JSX part. Here our injected HOC works and all components including sidebars are built.
   - This step includes important part - MDX context, which allows to replace certain React components with others. This is done in `~/common/translate-markdown.js`.
5. Further babel/webpack processing.

First place, where we can intercept the process and scrap heading is custom remark plugin injected to loader from step 3. Heading data can be then injected to `meta` variable - taking advantage of our custom loader (step 2).

Another very useful place is `~/common/translate-markdown.js` where our heading tags (H2, H3, H4) can be replaced with whatever we want. This feature has already been used to generate permalinks to certain header _(see `<Permalink/>` component)_. It uses [Github slugger](https://github.com/Flet/github-slugger) to generate slugs which are placed as HTML anchors. The slugger instance was held with React context.

I took advantage of that slugger context and decorated it with my custom `HeadingManager` _(name can change)_ which is now responsible for managing Slugger instance, storing metadata scraped from MDX plugin and storing whole heading list.

First approach I tried was scraping header levels and titles in MDX plugin and there create whole heading tree Then, at JSX level, only adding slugs to them, when replacing `<Hx/>` tags with `<Permalink />`. The problem is that title strings is the only thing that we have at that time and they're not guaranteed to be unique. The only thing we can be sure of is their order - fortunately they are processed synchronously top-to-bottom everywhere (tested experimentally).
My current approach is to integrate heading tree with current Permalink system. Heading tree is generated along with Permalink components (place where slugs are created too) and the system is more consistent (one peramlink = one heading) - we don't need to check twice if heading was already processed. This is synchronous too (orders are equal) and more flexible, because we can use data available at permalink generation (slug, styling) along with MDX plugin generated metadata (nesting level, type) and other arbitrary set parameters (max nesting level etc). It also works great with awesome changes made to `app.json` page by @tommybru - it really made implementation of this much easier.

Displaying the data is pretty straightforward. I did copy-paste of left sidebar, created `<DocumentationSidebarRight/>` component _(name can change)_ and used my React context to gather the data. I want it to look similar to left sidebar. Needed also to modify some React component reference holdings to be able to control scroll position, and thus highlight actively displayed component. I optimized it not to re-render unnecessarily, however CSS styling is currently a spaghetti-code and is going to be reworked. Titles being inline code (e.g. method names) are shortened. 

</details>

Some pages (Notifications) require custom heading treating, heading syntax in some places is used in non-standard way. We should have possibility to arbitrary set if we want to display it in sidebar, it's nesting level, possible custom display name. I see two possible approaches:
- _(IMPLEMENTED)_ Use custom JSX components instead of markdown headings. See _"How to use/customize"_ section.
- Use MDX plugin to parse other heading properties. Possible usage syntax: `### Some long long heading {showInSidebar=true,nestingLevel=2,sidebarDisplayName="Long heading"}`. Pros: Minimal effort needed for editor to customize. Cons: It's not part of markdown specification. Less elegant way to find it when building sidebar structure. _This appeared less elegant in implementation, but can be implemented if needed._

#### Problems:

Although core functionality works as expected, there are still some problems:

- Limited support for complex headings - those which partially include e.g. `inline code`. I haven't found any in current docs, but they might appear in future.
- Doesn't work on mobile - I haven't figured out why clicking anchors scrolls to incorrect place. For now, my sidebar is hidden on mobile.
- ~Some headings are too long, we need to consider either word wrap or customizing them by page.~ Done. Text headings are word-wrapped, code ones are truncated with ellipsis (...) sign. Pages requiring special treating can be customized individually. 

### For future PR's

This is growing big, I think splitting it is a good idea. All of these below could be done in parallel after this PR is merged:
- Adapt pages with too long headings, especially notifications.
- Add support for mobile site (see problems section above).
- ~Auto scroll sidebar when scrolling content to make active item always visible.~ _(done, #10230)_
- ~Add smooth scrolling for both sidebar and content.~ _(done, #10230)_

## Test Plan

After everything is done, I'll test the following:
- [x] Manual tests - visual style, behavior on complicated pages
- [x] `yarn test` - `jest` tests
- [x] `yarn build` - if builds successfully
- [x] `next export` - test exported website

### CODE REVIEW

**Comment by brentvatne** [comment]
thanks for the detailed writeup!

the only thing here that makes me a bit uneasy is that HeadingManager stores its state outside of React state and then that state is used in the sidebar render function. i'd prefer to have any state used in a render function belong to React, to avoid the risk of data going out of sync

**Comment by barthap** [comment]
This state is already managed by React. My `HeadingManager` instance is held in React context. The _GithubSlugger_ was held the same way till now. My manager is kinda decorator for that slugger - it just stores another data in addition to slugs.
Moreover, that context data, once stored when processing MDX, never changes at runtime.

File: docs/components/page-higher-order/withDocumentationElements.js [hunk 1]
```diff
@@ -1,28 +1,29 @@
-import * as React from 'react';
+import { MDXProvider } from '@mdx-js/tag';
```
Comment (byCedric) [inline docs/components/page-higher-order/withDocumentationElements.js:1]: As of MDX 1, this should be `@mdx-js/react` 😄 

You should be able to change this without breaking changes!

Comment (barthap) [reply docs/components/page-higher-order/withDocumentationElements.js:1]: Yeah, I know, I need to sync it with Your PR some way 😁

File: docs/next.config.js [hunk 2]
```diff
@@ -27,7 +28,14 @@ module.exports = withCSS({
     };
     config.module.rules.push({
       test: /.mdx?$/, // load both .md and .mdx files
-      use: [babelMdxLoader, '@mdx-js/loader', join(__dirname, './common/md-loader')],
+      use: [
+        babelMdxLoader,
+        {
+          loader: '@mdx-js/loader',
+          options: { mdPlugins: [headings] },
```
Comment (byCedric) [inline docs/next.config.js:35]: Also the `mdPlugins` are deprecated in v1, I think you can use `remarkPlugins`. But I'm not sure if the plugin is using the rehype or remark AST.

Comment (barthap) [reply docs/next.config.js:35]: This is the thing I also mentioned on your PR - it was badly documented on their docs. When you merge your MDX update PR, I'll fix it ASAP 😉

File: docs/common/headingManager.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
+import * as React from 'react';
+
+import * as Utilities from '~/common/utilities';
+
+/**
+ * These types directly correspond to MDAST node types
+ */
+export const HeadingType = {
+  Text: 'text',
+  Code: 'inlineCode',
```
Comment (bbarthec) [inline docs/common/headingManager.js:10]:
```suggestion
  InlineCode: 'inlineCode',
```
Why not keep the original name here? 🤔

File: docs/common/headingManager.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+   * Creates heading object instance and stores it
+   * @param {string | Object} title Heading display title or `<code/>` element
+   * @param {number|undefined} nestingLevel Override metadata heading nesting level.
+   * @param {*} additionalProps Additional properties passed to heading component
+   * @returns {Object} Newly created heading instance
+   */
+  addHeading(title, nestingLevel, additionalProps) {
+    // NOTE (barthap): workaround for complex titles containing both normal text and inline code
+    // changing this needs also change in `headingsMdPlugin.js` to make metadata loading correctly
+    if (Array.isArray(title)) title = title[0];
```
Comment (bbarthec) [inline docs/common/headingManager.js:58]:
```suggestion
    if (Array.isArray(title)) {
      title = title[0];
    }
```
```suggestion
    title = Array.isArray(title) ? title[0] : title;
```

File: docs/common/headingManager.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+   * Finds MDX-plugin metadata for specified title. Once found, it's marked as processed
+   * and will not be returned again.
+   * @param {string} realTitle Title to find metadata for
+   */
+  _findMetaForTitle(realTitle) {
+    const entry = this.meta.headings.find(
+      heading => heading.title === realTitle && !heading.processed
+    );
+    if (!entry) {
+      return undefined;
```
Comment (bbarthec) [inline docs/common/headingManager.js:96]:
```suggestion
      return;
```

File: docs/common/headingManager.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    return entry;
+  }
+
+  /**
+   * Checks if header title is an inline code block.
+   * @param {any} title Heading object to check
+   * @returns {boolean} true if header is a code block
+   */
+  _isCode(title) {
+    if (!title.props) return false;
```
Comment (bbarthec) [inline docs/common/headingManager.js:108]:
```suggestion
    if (!title.props) {
      return false;
    }
```

File: docs/components/DocumentationSidebarRight.js [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    maxNestingDepth: 4,
+  };
+
+  state = {
+    activeSlug: null,
+  };
+
+  handleContentScroll(contentScrollPosition) {
+    const { headingManager } = this.props;
+    const { headings } = headingManager;
```
Comment (bbarthec) [inline docs/components/DocumentationSidebarRight.js:56]:
```suggestion
    const { headings } = this.props.headingManager;
```

File: docs/components/DocumentationSidebarRight.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+      }
+      const { current } = heading.ref;
+      if (current && current.offsetTop >= contentScrollPosition) {
+        if (heading.slug !== this.state.activeSlug) {
+          this.setState({ activeSlug: heading.slug });
+        }
+        break;
+      }
+    }
+  }
```
Comment (bbarthec) [inline docs/components/DocumentationSidebarRight.js:70]:
```suggestion
    for (const { ref, slug } of headings) {
      if (!ref || !ref.current) {
        continue;
      }
      if (ref.current.offsetTop >= contentScrollPosition) {
        this.setState({ activeSlug: slug });
        return;
      }
    }
  }
```
React is not performing re-rendering if the state is not changed. And passing the same (sub-)state via `this.setState` is no-op in this case.

**Review by bbarthec — COMMENTED** [review]
Very neat! 🎉 
I can't wait to see this shipped to our docs 😊

File: docs/components/DocumentationSidebarRightLink.js [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+  const isCode = type === HeadingType.InlineCode;
+  const displayTitle = shortenCode && isCode ? trimCodedTitle(title) : title;
+  const linkFont = isCode ? Constants.fontFamilies.mono : undefined;
+  const fontSize = isCode ? 14 : undefined;
+
+  return (
+    <div className={STYLES_ACTIVE_CONTAINER}>
+      {isActive && <div className={STYLES_ACTIVE_BULLET} />}
+      <a
+        style={{ paddingLeft, fontFamily: linkFont, fontSize }}
```
Comment (barthap) [inline docs/components/DocumentationSidebarRightLink.js:105]:
```js
        style={{ paddingLeft, fontFamily: linkFont, fontSize }}
```
Something weird happens to this style when refreshing dev a few times. Even when my `paddingLeft` variable is always correct, inline style has it reseted to 0. 🙄 

Fortunately, this happens only in _dev_ and exported HTML is always correct.

**Review by bbarthec — APPROVED** [review]
We've discussed this PR a little with @byCedric and @esamelson and we want it to be released alongside the new SDK, namely early next week.
@byCedric do you think we might want to release it earlier, like today?
Do we have some sort of `staging` for documentation? 🤔

**Comment by byCedric** [comment]
Unfortunately, we don't have a staging environment for docs. But when it's ready, I can test it locally to see if it might cause some issue somewhere.

**Comment by bbarthec** [comment]
@byCedric, I believe it is ready already 😉 
When you have time take a look how it works locally, please 💪 👀

**Comment by byCedric** [comment]
> @byCedric, I believe it is ready already 😉
> When you have time take a look on how it works locally, please 💪 👀

@bbarthec great! I think I'm a bit confused about the various PRs but I approved #10230 and chatted with @barthap about it. So big 👍  and ❤️  from me!

**Comment by barthap** [comment]
I'll close this one to avoid further confusions 😅

**Pull request closed without merging**